lib/puppet/pops/types/enumeration.rb

```diff
@@ -1,33 +1,72 @@
+  #
+  # @param caller [Function] The function that needs to coerce its argument to an enumeration
+  # @param o [Object] The object to produce an `Enumerator` for
+  # @return [Enumerator] The produced `Enumerator` or nil if it couldn't be produced
+  # @raise [ArgumentError] In case an `Enumerator` cannot be produced
+  # @api public
+  def self.asserted_enumerable(caller, o)
+    enum = self.enumerator(o)
+    raise ArgumentError, "#{caller.class.name}(): wrong argument type (#{o.class.name}; must be something enumerable." if enum.nil?
+    enum
+  end
+
+  # Produces an `Enumerator` for `String`, `Integer`, `PIntegerType` or anything that respond to `each` by
+  # returning an `Enumerator`. The argument is returned in case it already is an `Enumerator`.
   #
+  # @param o [Object] The object to produce an `Enumerator` for
+  # @return [Enumerator,nil] The produced `Enumerator` or `nil` if it couldn't be produced
+  # @api public
   def self.enumerator(o)
     @@singleton ||= new
     @@singleton.enumerator(o)
   end
 
-  # Produces an Enumerator for Array, Hash, Integer, Integer Range, and String.
+  # Produces an `Enumerator` for `String`, `Integer`, `Range`, `PIntegerType` or anything that respond to `each` by
+  # returning an `Enumerator`. The argument is returned in case it already is an `Enumerator` or an `Iterable`.
   #
+  # @param o [Object] The object to produce an `Enumerator` for
+  # @return [Enumerator,nil] The produced `Enumerator` or `nil` if it couldn't be produced
+  # @api public
   def enumerator(o)
     case o
     when String
-      x = o.chars
-      # Ruby 1.9.3 returns an Enumerator, and 2.0.0 an Array
-      x.is_a?(Array) ? x.each : x
-    when Integer
-      o.times
+      EnumerationIterable.new(PStringType::DEFAULT, o.each_char)
     when Array
-      o.each
-    when Hash
-      o.each
-    when Puppet::Pops::Types::PIntegerType
-      # Not enumerable if representing an infinite range
-      o.enumerable? ? o.each : nil
+      EnumerationIterable.new(TypeCalculator.singleton.infer_and_reduce_type(o), o.each)
+    when Integer
+      o > 0 ? IntegerRangeIterator.new(PIntegerType.new(0, o - 1)) : nil
+    when PIntegerType
+      o.enumerable? ? IntegerRangeIterator.new(o) : nil
+    when PEnumType
+      EnumerationIterable.new(PStringType::DEFAULT, o.values)
+    when Range
+      min = o.min
+      if min.is_a?(Integer)
+        max = o.max
+        max > min ? IntegerRangeIterator.new(PIntegerType.new(min, max)) : nil
+      else
+        # A generalized form of the inferred type for the 'min' is used here since inferring the full
+        # range might waste a lot of memory.
+        EnumerationIterable.new(TypeCalculator.singleton.infer(min).generalize, o.each)
+      end
+    when Enumerator, Iterable
+      o
     else
-      nil
+      begin
+        result = o.each
+        result.is_a?(Enumerator) ? result : nil
```

Comment:
Hm, we have no type information here, requires you to accept `Iterable[Any]`. Do we really need this very open ended solution? Can we require a richer API? As an example, if we allow producer functions, and we have return types, we do know the produced type. Perhaps have either another method enumerator(o, T) that you need to use if what you created does not have type information. (That implies that the values that will be emitted needs to be type checked before yielded).